internal/cephfs/core/quiesce.go

```diff
@@ -0,0 +1,237 @@
+		state = Released
+	case "QUIESCING":
+		state = Quiescing
+	case "QUIESCED":
+		state = Quiesced
+	default:
+		state = QuiesceStates(set.Name)
+	}
+
+	return state
+}
+
+type FSQuiesceClient interface {
+	// FSQuiesce quiesces the subvolumes in the filesystem.
+	FSQuiesce(
+		ctx context.Context,
+		reserveName string,
+	) (*admin.QuiesceInfo, error)
+	// GetVolumes returns the list of volumes in the filesystem that are to be
+	// quiesced.
+	GetVolumes() []Volume
+	// FSQuiesceWithExpireTimeout quiesces the subvolumes in the
+	// filesystem with an expiration timeout.
+	FSQuiesceWithExpireTimeout(ctx context.Context,
+		reserveName string,
+	) (*admin.QuiesceInfo, error)
+	// ResetFSQuiesce resets the quiesce timeout for the subvolumes in
+	// the filesystem.
+	ResetFSQuiesce(ctx context.Context,
+		reserveName string,
+	) (*admin.QuiesceInfo, error)
+	// ReleaseFSQuiesce releases the quiesce on the subvolumes in the
+	// filesystem.
+	ReleaseFSQuiesce(ctx context.Context,
+		reserveName string,
+	) (*admin.QuiesceInfo, error)
+}
+
+type Volume struct {
+	VolumeID  string
+	ClusterID string
+}
+
+type fsQuiesce struct {
+	fsName  string
+	volumes []Volume
+	// subVolumeGroupMapping is a map of subvolumes to groups.
+	subVolumeGroupMapping map[string][]string
+	fsa                   *admin.FSAdmin
+}
+
+// NewFSQuiesce returns a new instance of fsQuiesce. It
+// take the filesystem name, the list of volumes to be quiesced, the mapping of
+// subvolumes to groups and the cluster connection as input.
+func NewFSQuiesce(
+	fsName string,
+	volumes []Volume,
+	mapping map[string][]string,
+	conn *util.ClusterConnection,
+) (FSQuiesceClient, error) {
+	fsa, err := conn.GetFSAdmin()
+	if err != nil {
+		return nil, err
+	}
+
+	return &fsQuiesce{
+		fsName:                fsName,
+		volumes:               volumes,
+		subVolumeGroupMapping: mapping,
+		fsa:                   fsa,
+	}, nil
+}
+
+// GetVolumes returns the list of volumes in the filesystem that are to be
+// quiesced.
+func (fq *fsQuiesce) GetVolumes() []Volume {
+	return fq.volumes
+}
+
+// getMembers returns the list of names in the format
+// group/subvolume that are to be quiesced. This is the format that the
+// ceph fs quiesce expects.
+// Example: ["group1/subvolume1", "group1/subvolume2", "group2/subvolume1"].
+func (fq *fsQuiesce) getMembers() []string {
+	volName := []string{}
+	for svg, sb := range fq.subVolumeGroupMapping {
+		for _, s := range sb {
+			name := svg + "/" + s
+			volName = append(volName, name)
+		}
+	}
+
+	return volName
+}
+
+func (fq *fsQuiesce) FSQuiesce(
+	ctx context.Context,
+	reserveName string,
+) (*admin.QuiesceInfo, error) {
+	opt := &admin.FSQuiesceOptions{
+		Timeout:    180,
+		AwaitFor:   0,
+		Expiration: 180,
+	}
+	log.DebugLog(ctx,
+		"FSQuiesce for reserveName %s: members:%v options:%v",
+		reserveName,
+		fq.getMembers(),
+		opt)
+	resp, err := fq.fsa.FSQuiesce(fq.fsName, admin.NoGroup, fq.getMembers(), reserveName, opt)
+	if resp != nil {
```

Comment:
We usually write code that returns early when an error happened and continues on success. Maybe consider changing the style here to what we do? Not sure if it is in the coding guidelines, it is not a blocker for me.